sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -452,46 +452,68 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
 
     val localIdx = ctx.freshName("localIdx")
     val localEnd = ctx.freshName("localEnd")
-    val range = ctx.freshName("range")
     val shouldStop = if (parent.needStopCheck) {
-      s"if (shouldStop()) { $number = $value + ${step}L; return; }"
+      s"if (shouldStop()) { $nextIndex = $value + ${step}L; return; }"
```

Comment:
in this case we are not very accurate in the metrics right? I mean we always say that we are returning a full batch, even though we have consumed less rows than a batch.

What about updating the metrics before returning? Something like `$inputMetrics.incRecordsRead($localIdx - $localEnd);`?

Reply:
BTW I do have a local branch that fixed this problem, I just don't have time to benchmark it yet. I'll send it out later and let's move the discussion there.

Reply:
I am not sure why you need a benchmark for this (unless you did something different from what I have suggested in the earlier comment). In that case it is a single metric update which happens only when stopping, it shouldn't introduce any significant overhead. Am I missing something? Anyway let's move the discussion to the next PR then, thanks.

Reply:
> Something like `$inputMetrics.incRecordsRead($localIdx - $localEnd);`?

`localIdx` is purely local to the loop, if we access it outside of the loop, we need to define `localIdx` outside of loop as well. This may have some performance penalty. cc @kiszk

Reply:
but `shouldStop` is called local to the loop, isn't it?

Reply:
`shouldStop` is called local, but metrics updating is not.

Anyway, JVM JIT is mysterious and we need to be super careful when updating this kind of hot loops. That said, I'm not confident of any changes to the hot loop without a benchmark.

Reply:
ok, let's get back to this eventually later, this is anyway not worse than before.

Reply:
Sorry for late comment. It would be good to discuss detail in another PR.

At first, I agree with necessary of benchmarking. 
1. I think that `localIdx` can be defined as local variable outside of the loop. Or, how about storing `localIdx` to another local variable only if `parent.needStopCheck` is `true`.
1. Since `shouldStop()` is simply without updating, we expect the JIT applies inlining and some optimizations.
1. If we want to call `incRecordRead`, it would be good to exit a loop using `break` and then call `incRecordRead`.